services/context-box/server/domain/usecases/enqueue_configs.go

```diff
@@ -0,0 +1,169 @@
+package usecases
+
+import (
+	"github.com/rs/zerolog/log"
+
+	"github.com/berops/claudie/proto/pb"
+	outboundAdapters "github.com/berops/claudie/services/context-box/server/adapters/outbound"
+	"github.com/berops/claudie/services/context-box/server/domain/ports"
+	"github.com/berops/claudie/services/context-box/server/utils"
+)
+
+// ConfigInfo is a data structure which holds data that context-box needs in order to function properly.
+// This data is required by EnqueueConfigs to decide which configs should be enqueued and which should not
+type ConfigInfo struct {
+	Name         string
+	MsChecksum   []byte
+	CsChecksum   []byte
+	DsChecksum   []byte
+	BuilderTTL   int32
+	SchedulerTTL int32
+	State        map[string]outboundAdapters.Workflow
+}
+
+// GetName function is required by the queue to evaluate equivalence
+func (c *ConfigInfo) GetName() string {
+	return c.Name
+}
+
+// HasError returns true if any cluster errored while building.
+func (c *ConfigInfo) HasError() bool {
+	for _, v := range c.State {
+		if v.Status == pb.Workflow_ERROR.String() {
+			return true
+		}
+	}
+
+	return false
+}
+
+var (
+	// Used for logging purposes
+	// Logs are generated whenever elements are added/removed to/from the queue
+	elementNameListForBuilderQueueWithPreviousState   []string
+	elementNameListForSchedulerQueueWithPreviousState []string
+)
+
+const (
+	defaultBuilderTTL   = 360
+	defaultSchedulerTTL = 5
+)
+
+// EnqueueConfigs checks all configs, decides if they should be enqueued or not and updates their TTLs
+func (u *Usecases) EnqueueConfigs() error {
+	configInfos, err := getConfigInfosFromDB(u.DB)
+	if err != nil {
+		return err
+	}
+
+	for _, configInfo := range configInfos {
+		// if item is already in some queue (scheduler / builder) then skip and move to the next item
+		if u.builderQueue.Contains(configInfo) || u.schedulerQueue.Contains(configInfo) {
+			continue
+		}
+
+		// Initially when the config is received from the frontend microservice, the desired state of the config is not built,
+		// due to which the DsChecksum will be nil.
+		if !utils.CompareChecksum(configInfo.DsChecksum, configInfo.MsChecksum) {
+			// If scheduler TTL is <= 0 AND config has no errorMessage, add item to the scheduler queue
+			if configInfo.SchedulerTTL <= 0 && !configInfo.HasError() {
+				if err := u.DB.UpdateSchedulerTTL(configInfo.Name, defaultSchedulerTTL); err != nil {
+					return err
+				}
+
+				// The item is put in the scheduler queue. The scheduler microservice will eventually pull the corresponding
+				// config and build its desired state
+				u.schedulerQueue.Enqueue(configInfo)
+				configInfo.SchedulerTTL = defaultSchedulerTTL
+
+				continue
+			}
+
+			// If the item is already present in the scheduler queue but the config is still not pulled by the scheduler
+			// microservice, then reduce its scheduler TTL by 1.
+			configInfo.SchedulerTTL = configInfo.SchedulerTTL - 1
```

Comment:
Here, we should add a check whether an error message is present. If there is, we do not want to decrease the TTL.

Reply:
Here is the original code -

```go
if !checksum.Equals(config.DsChecksum, config.MsChecksum) {
			// if scheduler ttl is 0 or smaller AND config has no errorMessage, add to scheduler Q
			if config.SchedulerTTL <= 0 && !config.HasError() {
				if err := database.UpdateSchedulerTTL(config.Name, defaultSchedulerTTL); err != nil {
					return err
				}
				queueScheduler.Enqueue(config)
				config.SchedulerTTL = defaultSchedulerTTL
				continue
			} else {
				config.SchedulerTTL = config.SchedulerTTL - 1
			}
		}
```

Does it make any difference whether we put the else statement or not?

Reply:
Well, the original code had the same issue. There is no reason why we cannot fix it now.
```go
if !checksum.Equals(config.DsChecksum, config.MsChecksum) {
			// if scheduler ttl is 0 or smaller AND config has no errorMessage, add to scheduler Q
			if config.SchedulerTTL <= 0 && !config.HasError() {
				if err := database.UpdateSchedulerTTL(config.Name, defaultSchedulerTTL); err != nil {
					return err
				}
				queueScheduler.Enqueue(config)
				config.SchedulerTTL = defaultSchedulerTTL
				continue
			} else !config.HasError() {
                                // Decrease the TTL only when config does not have any error
				config.SchedulerTTL = config.SchedulerTTL - 1
			}
		}
```
Please double-check that this would be sufficient.